### PR TITLE
i#3544 RV64: Implemented exit_cti_reaches_target and patch_branch

### PR DESCRIPTION
**Pull request by ksco**
Implemented `exit_cti_reaches_target()` and `patch_branch()`, which are required for execution of a simple program.

Issue: #3544

### CODE REVIEW

File: core/arch/riscv64/emit_utils.c [hunk 1]
```diff
@@ -82,8 +82,22 @@ bool
...
 {
-    /* FIXME i#3544: Not implemented */
-    ASSERT_NOT_IMPLEMENTED(false);
+    cache_pc branch_pc = EXIT_CTI_PC(f, l);
+    /* Compute offset as unsigned, modulo arithmetic. */
+    ptr_uint_t off = (ptr_uint_t)target_pc - (ptr_uint_t)branch_pc;
+    uint enc = *(uint *)branch_pc;
+    ASSERT(ALIGNED(branch_pc, 2) && ALIGNED(target_pc, 2));
+
+    if ((enc & 0x7f) == 0b1100011 &&
```
Comment (abhinav92003) [inline core/arch/riscv64/emit_utils.c:91]: For consistency, represent 0b1100011 in hex too.

File: core/arch/riscv64/emit_utils.c [hunk 2]
```diff
@@ -115,8 +129,29 @@ void
...
+    uint *pc_writable = (uint *)vmcode_get_writable_addr(branch_pc);
+    uint enc = *pc_writable;
+    ASSERT(ALIGNED(branch_pc, 2) && ALIGNED(target_pc, 2));
+    if ((enc & 0x7f) == 0b1100011 &&
+        /* BEQ, BNE */
+        (((enc >> 12) & 0x7) <= 0x1 ||
+         /* BLT, BGE, BLTU, BGEU */
+         ((enc >> 12) & 0x7) >= 0x4)) {
+        ASSERT(((off << (64 - 13)) >> (64 - 13)) == off);
+        *pc_writable = (enc & 0b00000001111111111111000001111111) |
```
Comment (abhinav92003) [inline core/arch/riscv64/emit_utils.c:143]: Please represent long offsets as hex.

File: core/arch/riscv64/emit_utils.c [hunk 3]
```diff
@@ -115,8 +129,29 @@ void
...
+            (((off >> 12) & 1) << 31) | (((off >> 5) & 63) << 25) |
```
Comment (abhinav92003) [inline core/arch/riscv64/emit_utils.c:144]: I don't understand this. Why are we sending different bits of the offset to different parts of the encoding? Maybe add a comment to explain.

Comment (ksco) [reply core/arch/riscv64/emit_utils.c:144]: RISC-V has a fragmented immediate encoding, I'll explain it with a comment. Is it ok to use something like what we did in the codec.c file?

```
/* Encode the immediate field of the B-type format as a pc-relative offset:
 * |  31   |30     25|24   20|19   15|14    12|11     8|   7   |6      0|
 * |imm[12]|imm[10:5]|  rs2  |  rs1  | funct3 |imm[4:1]|imm[11]| opcode |
 *  ^---------------^                          ^--------------^
 * From:
 * |31       12|  11   |10        5|4        1| 0 |
 * |  imm[12]  |imm[11]| imm[10:5] | imm[4:1] | 0 |
 */
```

Comment (ksco) [reply core/arch/riscv64/emit_utils.c]: Done in https://github.com/DynamoRIO/dynamorio/pull/6224/commits/d25a0f1e36f41879a5b854997500fc61defd0f5a, also fixed a bug.

File: core/arch/riscv64/emit_utils.c [hunk 4]
```diff
@@ -115,8 +129,41 @@ void
 patch_branch(dr_isa_mode_t isa_mode, cache_pc branch_pc, cache_pc target_pc,
              bool hot_patch)
 {
-    /* FIXME i#3544: Not implemented */
-    ASSERT_NOT_IMPLEMENTED(false);
+    /* Compute offset as unsigned, modulo arithmetic. */
+    ptr_int_t off = (ptr_uint_t)target_pc - (ptr_uint_t)branch_pc;
```
Comment (abhinav92003) [inline core/arch/riscv64/emit_utils.c:133]: Perhaps add an assert to make sure that the off is smaller than what can be handled by the branch instruction.

**Review by abhinav92003 — APPROVED** [review]

File: core/arch/riscv64/emit_utils.c (hunk 4 above)
Comment (ksco) [reply core/arch/riscv64/emit_utils.c:133]: It's asserted at line 142 and line 153. J-type and B-type have different reachability, so it's separated.

**Pull request merged**